listenbrainz_spark/recommendations/train_models.py

```diff
@@ -105,16 +194,136 @@ def train(training_data, validation_data, num_validation, ranks, lambdas, iterat
-        if best_model is None or validation_rmse < best_model.error:
-            best_model = Model(model=model, error=validation_rmse, rank=rank, lmbda=lmbda, iteration=iteration,
+        model_metadata.append((model_id, mt, rank, '{:.1f}'.format(lmbda), iteration, round(validation_rmse, 2), vt))
+        if best_model is None or validation_rmse < best_model.validation_rmse:
+            best_model = Model(model=model, validation_rmse=validation_rmse, rank=rank, lmbda=lmbda, iteration=iteration,
                 model_id=model_id, training_time=mt, rmse_time=vt)
 
-    best_model_metadata = {'error': '{:.2f}'.format(best_model.error), 'rank': best_model.rank, 'lmbda':
+    best_model_metadata = {'validation_rmse': best_model.validation_rmse, 'rank': best_model.rank, 'lmbda':
             best_model.lmbda, 'iteration': best_model.iteration, 'model_id': best_model.model_id, 'training_time':
-                best_model.training_time, 'rmse_time': best_model.rmse_time}
+                best_model.training_time, 'rmse_time': best_model.rmse_time, 'alpha': alpha}
     return best_model, model_metadata, best_model_metadata
 
+def save_model(model_id, model):
+    """ Save best model to HDFS.
+
+        Args:
+            model_id (str): Model identification string of best model.
+            model: Best model.
+    """
+    try:
+        model.save(listenbrainz_spark.context, config.HDFS_CLUSTER_URI + path.DATA_DIR + '/' + model_id)
+    except Py4JJavaError as err:
+        current_app.logger.error('Unable to save best model "{}"\n{}. Aborting...'.format(model_id,
+            str(err.java_exception)), exc_info=True)
+        sys.exit(-1)
+
+def save_model_index(test_rmse, model_id):
+    """ Overwrite model_index dataframe.
+        Refer to listenbrainz_spark.schema.model_index_schema
+
+        Args:
+            test_rmse (float): test rmse of best model.
+            model_id (str): Model identification string of best model.
+    """
+    metadata_row = schema.convert_model_index_to_row(test_rmse, model_id)
+    try:
+        # Create dataframe from the row object.
+        model_index_df = utils.create_dataframe(metadata_row, schema.model_index_schema)
+    except DataFrameNotCreatedException as err:
+        current_app.logger.error(str(err), exc_info=True)
+        sys.exit(-1)
+    try:
+        # The dataframe is overwritten since we wish to store model id of recently trained model.
+        utils.save_parquet(model_index_df, path.INDEX)
+    except DataFrameNotAppendedException as err:
+        current_app.logger.error(str(err), exc_info=True)
+        sys.exit(-1)
+
+def delete_model(model_id):
+    """ Delete a model.
+
+        Args:
+            model_id (str): Model identification string of model to be deleted.
+    """
+    utils.delete_dir(path.DATA_DIR + model_id, recursive=True)
+
+def save_model_delete_metadata(model_id, timestamp):
```

Comment:
what? this is very confusing and the docstring doesn't really help at all.

Reply:
I agree. Thanks for pointing out :)

Reply:
how about save_deleted_model_ids?